ansible/roles/set-repositories/tasks/rhn-repos.yml

```diff
@@ -1,41 +1,31 @@
 # vim: set ft=ansible:
 ---
+
 - name: Force unregister before register
-  redhat_subscription:
-    state: absent
-  register: task_result
-  until: task_result | succeeded
-  retries: 10
-  delay: 5
+  command: "{{ item }}"
+  args:
+    warn: False
+  with_items:
+  - 'subscription-manager clean'
+  - 'subscription-manager remove --all'
+  - 'yum remove -y "katello-ca-consumer-*"'
 
-- name: register node with subscription-manager
+- name: 'Register system using Red Hat Subscription Manager'
   redhat_subscription:
     state: present
     username: "{{ rhel_subscription_user }}"
     password: "{{ rhel_subscription_pass }}"
-    autosubscribe: false
-  register: task_result
-  until: task_result | succeeded
-  retries: 10
-  delay: 5
+    pool_ids: "{{ rhsm_pool_ids }}"
+    auto_attach: false
 
-# TODO: should take pool ids from a var
-- name: get product pool id
-  shell: /usr/bin/subscription-manager list --all --available --matches="*{{rhn_pool_id_string}}*" | awk '/Pool ID/ {print $3}' | head -1
-  # command: subscription-manager list --all --available --matches="OpenShift Container Platform" | awk '/Pool ID/ {print $3}' | head -1
-  register: pool_id
-  until: pool_id | succeeded
-  retries: 10
-  delay: 5
-
-- name: attach host to subscription pool
-  shell: /usr/bin/subscription-manager attach --pool={{ pool_id.stdout }}
-  register: task_result
-  until: task_result.rc == 0
-  retries: 10
-  delay: 5
-
-- name: enable repos for rhel
-  command: subscription-manager repos --enable "{{ item }}"
+- name: "Build command line for repos to enable"
+  set_fact:
+    repos_params: "{{ repos_params|default('') }} --enable={{ item }}"
   with_items:
-    - '{{ rhel_repos }}'
+  - "{{ rhel_repos }}"
+
+- name: "Run 'subscription-manager to disable/enable repos"
+  command: "subscription-manager repos {{ repos_params }}"
```

Comment:
you could probably just do:

```yaml
- name: "Run 'subscription-manager to disable/enable repos"
   command: >-
    subscription-manager repos {% for repo in rhel_repos %}
    --enable={{ repo }}
    {% endfor %}
  when: rhel_repos | default([]) | length > 0
```

?

Reply:
Yes, indeed we can for this purpose as we are assuming this is a new deployment and actually unregister the system in case the image used was already registered. Just changed that part of the code.

I think this is ok for this purpose, but if you think this role is going to be used for more than registering systems during deployment phase we could extend it to cover more use cases, but maybe this is too much here.

Check these examples (as I'm using both as a base for this) to see what I'm referring to:

- https://github.com/redhat-sap/sap-rhsm/blob/master/tasks/main.yml
- https://github.com/redhat-cop/infra-ansible/blob/master/roles/rhsm/tasks/main.yml